### PR TITLE
New docker image for Go+Docker+Heroku

### PR DESCRIPTION
**Pull request by philpennock**
We want the Docker tools, to talk to a remote Docker server, for use
with Circle CI's `setup_remote_docker` so that we can build Docker
images when running in Docker.  We also want Go, but now can just use
upstream Golang images, and we add Heroku as before, and our packages.

This image drops the user to a non-root runtime user, with no way back
to root.  Not a problem: if we need more privs, we can rebuild the
container.

### CODE REVIEW

File: docker-go/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# We don't use /go because we don't build as root and 777 permissions are daft
+RUN rm -rf /go
+
+# nb: we don't have a password and have not set up sudo, so no way back at this
+# point.  Do we _want_ to still have root?
+USER ${RUNTIME_USER}
+# Want to just remove the Go image's GOPATH setting but while we can replace
+# ENV in Docker, we can't unset.  So suck it up and just set it, old style.
+ENV GOPATH=/home/${RUNTIME_USER}/go
+ENV PATH ${GOPATH}/bin:/usr/local/go/bin:$PATH
```
Comment (cee-dub) [inline docker-go/Dockerfile]: Missing equals sign?

**Comment by cee-dub** [comment]
In general, this looks promising. Am.i correct that we'd use this image in CircleCI to run tests and to run the build of a different Docker image  that replaces slugs for production deployment?

File: docker-go/Dockerfile (hunk 1 above)
Comment (philpennock) [reply docker-go/Dockerfile]: Nope, although for consistency I should use one or the other.
https://docs.docker.com/engine/reference/builder/#env

**Comment by philpennock** [comment]
You are correct.  We'd replace the image we're using now, from the `circlego` sub-directory.

We basically need the Docker binary client available, on top of other stuff we have already added.  It looked like using the official golang images presented a marginally saner base than Circle's Golang one.

File: docker-go/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	&& curl -fsSL "https://cli-assets.heroku.com/heroku-cli/channels/stable/heroku-cli-linux-x64.tar.gz" -o heroku.tar.gz \
+	&& tar -zxf heroku.tar.gz \
+	&& mv heroku-cli-*-linux-x64 /usr/local/lib/heroku \
+	&& ln -s /usr/local/lib/heroku/bin/heroku /usr/local/bin/heroku \
+	&& rm heroku.tar.gz \
+	&& heroku version
+
+WORKDIR /home/${RUNTIME_USER}
+# We don't use /go because we don't build as root and 777 permissions are daft
+RUN rm -rf /go
```
Comment (cee-dub) [inline docker-go/Dockerfile]: Where does the go runtime live in this image? Obviously not in here.

File: docker-go/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+# Want to just remove the Go image's GOPATH setting but while we can replace
+# ENV in Docker, we can't unset.  So suck it up and just set it, old style.
+ENV GOPATH=/home/${RUNTIME_USER}/go
+ENV PATH=${GOPATH}/bin:/usr/local/go/bin:$PATH
+
+# Install our Go tools
+RUN go version && \
+	go get -u -v github.com/jstemmer/go-junit-report && \
+	go get -u -v github.com/nbio/slugger && \
+	go get -u -v github.com/nbio/cart && \
```
Comment (cee-dub) [inline docker-go/Dockerfile]: How do we keep these tools up to date in the image? Is doing this sufficient to update them each time the Dockerfile is executed?

Comment (philpennock) [reply docker-go/Dockerfile]: Yes: every time Dockerfile is run, it's effectively run in a clean filesystem.  The `-u` flag to `go get` does an Update.  So building the image will update these tools.

File: docker-go/Dockerfile (hunk 2 above)
Comment (philpennock) [reply docker-go/Dockerfile]: In `/usr/local/go`

The start of the file:
```
ARG GOLANG_BASE_IMAGE=1.9.3-stretch
FROM golang:${GOLANG_BASE_IMAGE}
```

So:

```
% docker run -it --rm golang:1.9.3-stretch
root@7aaa093586e5:/go# ls /go
bin  src
root@7aaa093586e5:/go# ls /usr/local/go
AUTHORS		 CONTRIBUTORS  PATENTS	  VERSION  bin	 doc	      lib   pkg		src
CONTRIBUTING.md  LICENSE       README.md  api	   blog  favicon.ico  misc  robots.txt	test
root@7aaa093586e5:/go#
```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,8 @@
 # CI
+
 CI support for Domainr
+
+Used to create: `domainr/docker-go:go1.9.4-docker17.12.0-ce` or the like
```
Comment (cee-dub) [inline README.md]: This name is configured in Docker hub, right? The config there will build `Dockerfile` at the root as `:latest`.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,8 @@
...
+
+Possibly used in future builds, having the Docker client tool to talk to a
+remote docker server, so should work with Circle CI's `setup_remote_docker`.
```
Comment (cee-dub) [inline README.md]: It's going to be used immediately. :D

**Review by cee-dub — APPROVED** [review]
Let's see if it blends!

File: README.md (hunk 4 above)
Comment (ydnar) [inline README.md]: Can we call this `domainr/ci`?

**Review by ydnar — APPROVED** [review]
LGTM, can we call this `domainr/ci` instead of the longer mangled image name? Does that matter?

File: README.md (hunk 4 above)
Comment (philpennock) [reply README.md]: Yes, that was "what I was doing locally" and in fact this statement is wrong.

**Comment by philpennock** [comment]
It's called whatever you tag it; auto-builds will tag it as `domainr/ci:latest` probably, depending upon what Eric set up.  That text refers to what I was doing before, missed another change.

File: README.md (hunk 5 above)
Comment (ydnar) [reply README.md]: Yep

**Comment by cee-dub** [comment]
Thumbs up!

**Pull request merged**